Haskell Parser Combinators/article.md

```diff
@@ -0,0 +1,1375 @@
+string [] = pure []
+string (x : xs) = (:) <$> char x <*> string xs
+```
+
+Or even in terms of `Monad`:
+
+```hs
+string [] = return []
+string (x : xs) = do
+  y <- char x
+  ys <- string xs
+  return (y : ys)
+```
+
+Let's try it in GHCi:
+
+```hs
+>>> runParser (string "Haskell") "Haskell"
+Right ("Haskell","")
+
+>>> runParser (string "Haskell") "Halloween"
+Left [Unexpected 'l']
+```
+
+Finally, our last instance: `Alternative`. This may be new for you, so I will explain. First, take a look at its definition, which can be found in `Control.Applicative`:
+
+```hs
+class Applicative f => Alternative f where
+  empty :: f a
+  (<|>) :: f a -> f a -> f a
+```
+
+I mentioned that `pure` acts as `1`, which is a parser that always succeeds without consuming input. `empty` works like the converse (or `0`), that is a parser that always fails without consuming input.
+
+For `(<|>)`, we want to implement it such that **the first parser to succeed is used**. That is, we first try the left operand, and failing that, we try the second one. The table below shall guide you:
+
+| |Left parser failed|Left parser succeeded|
+|-|-|-|
+|Right parser failed|Errors from both parsers|Left parser|
+|Right parser succeeded|Right parser|Left parser|
+
+Here is an implementation of `Alternative`. We constrain `i` and `e` with `Eq` so we can use `nub` to remove duplicated errors. Note that this is not the best or most efficient way to merge errors, but for our purposes, it's good enough.
+
+```hs
+instance (Eq i, Eq e) => Alternative (Parser i e) where
+  empty = Parser $ \_ -> Left [Empty]
+
+  Parser l <|> Parser r = Parser $ \input ->
+    case l input of
+      Left err ->
+        case r input of
+          Left err' -> Left $ nub $ err <> err'
+          Right (output, rest) -> Right (output, rest)
+      Right (output, rest) -> Right (output, rest)
+```
+
+Finally, let's test it!
+
+```hs
+>>> runParser (string "hello" <|> string "greetings") "hello, world"
+Right ("hello",", world")
+
+>>> runParser (string "hello" <|> string "greetings") "greetings, world"
+Right ("greetings",", world")
+
+>>> runParser (string "hello" <|> string "greetings") "bye, world"
+Left [Unexpected 'b']
+
+>>> runParser (empty <|> pure ()) ""
+Right ()
+
+>>> runParser (pure () <|> empty) ""
+Right ()
+```
+
+We can see our errors composing with a more complicated expression:
+
+```hs
+>>> runParser ((string "hello" *> string ", globe") <|> string "greetings") "hello, world"
+Left [Unexpected 'w',Unexpected 'h']
+```
+
+Hopefully, this should give you an intuition of how parser combinators work. Popular parser combinator libraries are more complex than this, as they tend to keep track of extra state for better error messages, optimization, information about line and column, etc., but the overall operations between them it similar.
+
+The complete code for the parser with exercises solutions can be found [here](https://gist.github.com/heitor-lassarote/3e7314956e86b8227f6f6040e69aca9d).
```

Comment:
There is something wrong here: 

1. exercises & solutions? 
2. I only see parser code in the gist.

Reply:
I reworded a bit, mentioned it contains the solutions to the exercises below.

> I only see parser code in the gist.

I mentioned it's only for this section, which only has the parser code. Or am I missing something?

Reply:
I meant that it doesn't have the solutions to the exercises, at least it seems that way to me.

Reply:
My bad, I thought it didn't have the solutions, but it does.